### PR TITLE
New style Gaussian

### PR DESCRIPTION
**Pull request by hvy**
New style gaussian noise to support double backprop.

### CODE REVIEW

**Comment by hvy** [comment]
Rebased on top of master.

File: chainer/functions/noise/gaussian.py [hunk 1]
```diff
@@ -28,23 +35,24 @@ def check_type_forward(self, in_types):
...
+        self.retain_inputs((1,))
 
-        cupy = cuda.cupy
         mean, ln_var = inputs
         if self.eps is None:
-            self.eps = cupy.random.standard_normal(
+            self.eps = cuda.cupy.random.standard_normal(
                 ln_var.shape, dtype=mean.dtype)
 
         self.noise = cuda.cupy.empty_like(mean)
```
Comment (takagi) [inline chainer/functions/noise/gaussian.py]: `self.noise` is enough to be a local variable now.

File: chainer/functions/noise/gaussian.py [hunk 2]
```diff
@@ -28,23 +35,24 @@ def check_type_forward(self, in_types):
...
         mean, ln_var = inputs
         if self.eps is None:
-            self.eps = numpy.random.standard_normal(ln_var.shape) \
-                                   .astype(numpy.float32)
+            self.eps = (
+                numpy.random.standard_normal(ln_var.shape)
+                .astype(numpy.float32)
+            )
 
         self.noise = numpy.exp(ln_var * mean.dtype.type(0.5)) * self.eps
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/noise/gaussian.py]: Changed to local variable.

File: chainer/functions/noise/gaussian.py (hunk 1 above)
Comment (hvy) [reply chainer/functions/noise/gaussian.py]: Ditto!

File: chainer/functions/noise/gaussian.py [hunk 3]
```diff
@@ -55,26 +63,38 @@ def forward_gpu(self, inputs):
...
+        mean (~chainer.Variable):
+            Input variable representing the mean :math:`\\mu`.
+        ln_var (~chainer.Variable):
+            Input variable representing the logarithm of a variance
+            :math:`\\log(\\sigma^2)`.
 
     Returns:
-        ~chainer.Variable: Output variable.
+        ~chainer.Variable:
+            Output variable with the shape of ``mean`` and/or ``ln_var``.
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/functions/noise/gaussian.py]: You are right, fixed.

File: tests/chainer_tests/functions_tests/noise_tests/test_gaussian.py [hunk 4]
```diff
@@ -39,9 +43,19 @@ def test_forward_gpu(self):
         self.check_forward(cuda.to_gpu(self.m), cuda.to_gpu(self.v))
 
     def check_backward(self, m_data, v_data, y_grad):
+        # Instantiate the FunctionNode object outside the function that is
+        # tested in order to reuse the same noise for the numerical gradient
+        # computations (the noise is generated once during its first forward
+        # pass, then reused)
+        # TODO(hvy): Do no expose interals of the tested function using
```
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/functions_tests/noise_tests/test_gaussian.py]: Removed this comment.

**Comment by hvy** [comment]
@takagi san, thanks for the review. I fixed the code accordingly. Please see comments!

**Comment by takagi** [comment]
Would you resolve the conflict?

**Comment by hvy** [comment]
Rebased!

**Pull request merged**